Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -88,79 +90,55 @@ + (instancetype)snapshotWithFirestore:(FIRFirestore *)firestore
 @end
 
 @implementation FIRDocumentSnapshot {
-  FIRSnapshotMetadata *_cachedMetadata;
-  DocumentKey _internalKey;
+  DocumentSnapshot _shim;
 }
 
-@dynamic metadata;
-
 - (instancetype)initWithFirestore:(FIRFirestore *)firestore
                       documentKey:(DocumentKey)documentKey
                          document:(nullable FSTDocument *)document
                         fromCache:(BOOL)fromCache
                  hasPendingWrites:(BOOL)pendingWrites {
   if (self = [super init]) {
-    _firestore = firestore;
-    _internalKey = std::move(documentKey);
-    _internalDocument = document;
-    _fromCache = fromCache;
-    _pendingWrites = pendingWrites;
+    _shim = DocumentSnapshot{firestore, std::move(documentKey), document, fromCache, pendingWrites};
```

Comment:
As discussed via chat, "shim" isn't quite the right concept here. A shim is something that might sit in between the Objective-C API and the C++ core API, adjusting parameters to fit. If anything the Objective-C wrapper is more of a shim than the the underlying DocumentSnapshot which is the actual implementation.

Ignoring Cocoa's use of the term, this is more like a "delegate". See for example, the way that term is used in e.g. [`ForwardingList`](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ForwardingList.java#L61).

Unfortunately, in Cocoa, a "delegate" has a [narrower meaning](https://developer.apple.com/library/archive/documentation/General/Conceptual/DevPedia-CocoaCore/Delegation.html).

In Guava's usage a delegate is a sort of underlying object that's supposed to do all the work. In Cocoa's usage, a delegate is a cooperating object that gets to receive notifications and specify policy without being a part of the implementation. While the former fits, the latter doesn't 😖.

In this particular case it doesn't happen, but in many others it seems as if we'll have multiple members that need to use (e.g. the firestore reference and something else). In those cases calling one the delegate "delegate" or "shim" and not the other one will be unclear.

All of which is just to say, how about we just avoid the general naming problem of what to call this kind of underlying object and just name this `_snapshot`?